onnx/shape_inference/implementation.h

```diff
@@ -0,0 +1,135 @@
+    return allInputTypes_.size();
+  }
+  const TypeProto_Tensor* getInputType(size_t index) const override {
+    return allInputTypes_[index];
+  }
+  size_t getNumOutputTypes() const override {
+    return allOutputTypes_.size();
+  }
+  TypeProto_Tensor* getOutputType(size_t index) override {
+    return &allOutputTypes_[index];
+  }
+  std::unordered_map<std::string, const AttributeProto *> attributesByName_;
+  std::vector<const TypeProto_Tensor*> allInputTypes_;
+  std::vector<TypeProto_Tensor> allOutputTypes_;
+};
+
+void InferShapes(ModelProto& m) {
+  auto* g = m.mutable_graph();
+
+  std::unordered_map<std::string, TypeProto_Tensor *> valueTypesByName;
+  for (auto& vi : *g->mutable_value_info()) {
+    valueTypesByName[vi.name()] = vi.mutable_type()->mutable_tensor_type();
+  }
+  for (auto& vi : *g->mutable_input()) {
+    valueTypesByName[vi.name()] = vi.mutable_type()->mutable_tensor_type();
+  }
+  for (auto& vi : *g->mutable_output()) {
+    valueTypesByName[vi.name()] = vi.mutable_type()->mutable_tensor_type();
+  }
+
+  for (const auto& n : g->node()) {
+    auto schema = OpSchemaRegistry::Schema(n.op_type());
+    if (schema) {
+      InferenceContextImpl ctx;
+
+      {
+        std::unordered_map<std::string, const AttributeProto *> attributesByName;
+        for (const auto& attr : n.attribute()) {
+          attributesByName[attr.name()] = &attr;
+        }
+        ctx.attributesByName_ = std::move(attributesByName);
+      }
+
+      {
+        std::vector<const TypeProto_Tensor *> allInputTypes;
+
+        for (const auto& input : n.input()) {
+          auto iter = valueTypesByName.find(input);
+          if (iter != valueTypesByName.end()) {
+            allInputTypes.push_back(iter->second);
+          } else {
+            allInputTypes.push_back(nullptr);
+          }
+        }
+
+        ctx.allInputTypes_ = std::move(allInputTypes);
+      }
+
+      ctx.allOutputTypes_.resize(n.output_size());
+
+      schema->GetShapeInferenceFunction()(ctx);
+
+      for (int i = 0; i < n.output_size(); i++) {
+        const auto& output = n.output(i);
+        const auto& inferredType = *ctx.getOutputType(i);
+
+        // In this case, we have no new information, so don't bother
+        // to add a contentless ValueInfo.
+        if (inferredType.elem_type() == TensorProto::UNDEFINED &&
+            !inferredType.has_shape()) {
+          continue;
+        }
+
+        // If there is already a ValueInfo associated with this
+        // output, reuse it. Otherwise add a new one.
+        auto iter = valueTypesByName.find(output);
+        TypeProto_Tensor* existingType = nullptr;
+        if (iter != valueTypesByName.end()) {
+          existingType = iter->second;
+        } else {
+          auto vi = g->add_value_info();
+          vi->set_name(output);
+          existingType = vi->mutable_type()->mutable_tensor_type();
+        }
+
+        // Incorporate the inferred information.
```

Comment:
maybe extract all this comparison semantics into a separate function?

Reply:
hmm